### PR TITLE
Change outcome of CannotDecide operations

### PR DESCRIPTION
**Pull request by lukaszachy**
Implements Alex's proposal.

### CODE REVIEW

**Comment by lukaszachy** [comment]
@t184256 Could you review please as well?

File: tests/unit/test_context.py [hunk 1]
```diff
@@ -479,17 +483,14 @@ def test_matches_groups(self):
...
-        )  # skip over 'foo' part since it is not defined
         assert not context.matches("foo=bar and distro = rhel")
-        assert not context.matches("foo=bar or distro = centos-8.9.0")
-
         # Whole rule cannot be decided
         for undecidable in [
             "foo = baz",
             "foo = baz and distro ~<= centos-7.2",  # both are CannotDecide
             "foo = baz and distro ~<= fedora-32 or do=done",
+            "foo=bar and distro = centos-8.2.0", # CannotDecide and True
```
Comment (t184256) [inline tests/unit/test_context.py:492]: `=` is not surrounded by whitespace, not sure if intended

File: tests/unit/test_context.py [hunk 2]
```diff
@@ -648,6 +649,37 @@ def test_known_troublemakers(self):
...
 
+    def test_cannotdecides(self):
+        # https://github.com/psss/fmf/issues/117
+        # CannotDecide and True = True and CannotDecide = CannotDecide
+        # CannotDecide and False = False and CannotDecide = False
+        # CannotDecide or True = True or CannotDecide = True
+        # CannotDecide or False = False or CannotDecide = CannotDecide
+        _true = "foo == bar"
+        _false = "foo != bar"
+        _cannot =  "baz == bar"
```
Comment (t184256) [inline tests/unit/test_context.py:660]: double space

**Review by t184256 — COMMENTED** [review]
Minor formatting nitpicks inline. Logic looks OK.

I'd also like to see a testcase OR'ing on a same axis with one or both values being CannotDecide, smth like `distro < fedora-33 or distro < centos-6.9` for `fedora-33`, `fedora-34`, `centos-6.5` and `centos-7.7`. This is going to be common and I'd like to have it explicit.

**Comment by lukaszachy** [comment]
> I'd also like to see a testcase OR'ing on a same axis with one or both values being CannotDecide, smth like `distro < fedora-33 or distro < centos-6.9` for `fedora-33`, `fedora-34`, `centos-6.5` and `centos-7.7`. This is going to be common and I'd like to have it explicit.

Extended `test_known_troublemakers` with proposed scenario.

**Review by t184256 — APPROVED** [review]
LGTM! I probably meant to ask for `fedora-32`, not `fedora-33`, but this should be enough as well.

File: tests/unit/test_context.py [hunk 3]
```diff
@@ -648,6 +649,47 @@ def test_known_troublemakers(self):
...
+        for a, op, b in [
+            (_cannot, 'and', _true),
+            (_true, 'and', _cannot),
+            (_cannot, 'or', _false),
+            (_false, 'or', _cannot),
+            ]:
+            exp = "{0} {1} {2}".format(a, op, b)
+            with pytest.raises(CannotDecide):
+                env.matches(exp)
+        for outcome, a, op ,b in [
```
Comment (psss) [inline tests/unit/test_context.py:681]:
```suggestion
        for outcome, a, op, b in [
```

File: docs/context.rst [hunk 4]
```diff
@@ -56,10 +56,20 @@ Lazy evaluation
...
+    Boolean operations with CannotDecide
+
+============  ===  ============  ============
+CannotDecide  and  True          CannotDecide
+CannotDecide  and  False         False
+CannotDecide  or   True          True
+CannotDecide  or   False         CannotDecide
+CannotDecide  and  CannotDecide  CannotDecide
+CannotDecide  or   CannotDecide  CannotDecide
+============  ===  ============  ============
```
Comment (psss) [inline docs/context.rst:72]: Looking at the table I'm missing the `==` before the last column. It's a bit confusing. What about just writing theses examples as expressions?

    CannotDecide and True == CannotDecide

File: docs/context.rst [hunk 5]
```diff
@@ -159,22 +169,22 @@ right side defines if the minor comparison is evaluated or not.
...
-centos-7     skip         skip    True
-centos-8.1   skip         True    False
-centos-8.2   skip         False   False
-centos-8     skip         skip    False
+centos-7.8   True        cannot    True
+centos-7.9   False       cannot    True
+centos-7     cannot      cannot    True
+centos-8.1   cannot      True      False
+centos-8.2   cannot      False     False
+centos-8     cannot      cannot    False
```
Comment (psss) [inline docs/context.rst:177]:
```suggestion
centos-8     CannotDecide   CannotDecide   False
```

Let's use the `CannotDecide` name consistently across the whole doc.

File: docs/context.rst [hunk 6]
```diff
@@ -123,7 +133,7 @@ right side match then the two values are equal::
 
 However, comparing order of two values is defined only if they
 match by name. If names don't match then values cannot be
-compared and the expression is skipped::
+compared and the expression has CannotDecide outcome::
```
Comment (psss) [inline docs/context.rst:136]:
```suggestion
compared and the expression has ``CannotDecide`` outcome::
```

**Review by psss — CHANGES_REQUESTED** [review]
Looks good, proposed just a couple of docs improvements.

**Review by psss — APPROVED** [review]
Looks good. Proposing to use sections to be able to link directly.

**Review by lukaszachy — COMMENTED** [review]
OK. Thanks

**Pull request merged**